python/sdk/merlin/pyfunc.py

```diff
@@ -369,3 +393,209 @@ def infer(self, model_input: pandas.DataFrame) -> Union[numpy.ndarray,
+        else:
+            pyfunc_base_image = (
+                f"ghcr.io/caraml-dev/merlin/merlin-pyfunc-base:{VERSION}"
+            )
+
+    dockerfile_path = copy_pyfunc_dockerfile(context_path)
+
+    _clone_merlin_repo(context_path)
+
+    image_tag = f"{model_name}-{model_version}"
+
+    print(f"Building Docker image {image_tag}")
+    if debug:
+        print("Context path:", context_path)
+        print("Dockerfile path:", dockerfile_path)
+        print("Dependencies path:", dependencies_path)
+        print("Model artifacts path:", artifact_path)
+
+    _build_image(
+        image_tag=image_tag,
+        context_path=context_path,
+        dependencies_path=dependencies_path,
+        artifact_path=artifact_path,
+        pyfunc_base_image=pyfunc_base_image,
+        dockerfile_path=dockerfile_path,
+        debug=debug,
+    )
+
+    print("Running PyFunc local server")
+    _run_container(
+        image_tag=image_tag,
+        model_name=model_name,
+        model_version=model_version,
+        model_full_name=f"{model_name}-{model_version}",
+        port=port,
+        env_vars=env_vars,
+        protocol=protocol,
+    )
+
+
+def _clone_merlin_repo(context_path: str):
+    repo_url = "https://github.com/caraml-dev/merlin.git"
+    repo_path = f"{context_path}/merlin"
+    if not os.path.isdir(repo_path):
+        repo = Repo.clone_from(repo_url, repo_path)
+        if "dev" in VERSION or "0.0.0" in VERSION:
+            repo.git.checkout("main")
+        else:
+            repo.git.checkout(f"v{VERSION}")
+
+
+def _build_image(
+    image_tag,
+    context_path,
+    dependencies_path,
+    artifact_path,
+    pyfunc_base_image,
+    dockerfile_path,
+    debug,
+):
+    docker_api_client = APIClient()
+    logs = docker_api_client.build(
+        path=context_path,
+        tag=image_tag,
+        buildargs={
+            "BASE_IMAGE": pyfunc_base_image,
+            "MODEL_DEPENDENCIES_URL": os.path.basename(dependencies_path),
+            "MODEL_ARTIFACTS_URL": artifact_path,
+        },
+        dockerfile=os.path.basename(dockerfile_path),
+        decode=True,
+    )
+
+    wait_build_complete(logs, debug)
+
+
+def _run_container(
+    image_tag,
+    model_name,
+    model_version,
+    model_full_name,
+    port,
+    env_vars: Dict[str, str] = None,
+    protocol: Protocol = None,
+):
+    docker_client = docker.from_env()
+
+    # Stop all previous containers to avoid port conflict
+    started_containers = docker_client.containers.list(
+        filters={"name": model_full_name}
+    )
+    for started_container in started_containers:
+        started_container.remove(force=True)
+
+    if env_vars is None:
+        env_vars = {}
+
+    env_vars["CARAML_HTTP_PORT"] = "8080"
+    env_vars["CARAML_GRPC_PORT"] = "9000"
+    env_vars["CARAML_MODEL_NAME"] = model_name
+    env_vars["CARAML_MODEL_VERSION"] = model_version
+    env_vars["CARAML_MODEL_FULL_NAME"] = model_full_name
+    env_vars["WORKERS"] = "1"
+
```

Comment:
missing 
`env_vars["CARAML_PROTOCOL"] = protocol.value` or this will always run HTTP server ([config](https://github.com/caraml-dev/merlin/blob/main/python/pyfunc-server/pyfuncserver/config.py#L81))

Reply:
thanks!